browser/ai_chat/code_execution_tool.cc

```diff
@@ -0,0 +1,207 @@
+}
+
+std::optional<base::Value::Dict> CodeExecutionTool::InputProperties() const {
+  return CreateInputProperties(
+      {{"script", StringProperty("The JavaScript code to execute")}});
+}
+
+std::optional<std::vector<std::string>> CodeExecutionTool::RequiredProperties()
+    const {
+  return std::vector<std::string>{"script"};
+}
+
+std::variant<bool, mojom::PermissionChallengePtr>
+CodeExecutionTool::RequiresUserInteractionBeforeHandling(
+    const mojom::ToolUseEvent& tool_use) const {
+  return false;
+}
+
+bool CodeExecutionTool::SupportsConversation(
+    bool is_temporary,
+    bool has_untrusted_content,
+    mojom::ConversationCapability conversation_capability) const {
+  // Support all conversation types for now
+  return true;
+}
+
+void CodeExecutionTool::SetExecutionTimeLimitForTesting(
+    base::TimeDelta time_limit) {
+  execution_time_limit_for_testing_ = time_limit;
+}
+
+void CodeExecutionTool::UseTool(const std::string& input_json,
+                                UseToolCallback callback) {
+  auto input_dict = base::JSONReader::ReadDict(
+      input_json, base::JSON_PARSE_CHROMIUM_EXTENSIONS);
+  if (!input_dict.has_value()) {
+    std::move(callback).Run(CreateContentBlocksForText(
+        "Error: Invalid JSON input, input must be a JSON object"));
+    return;
+  }
+
+  const std::string* script = input_dict->FindString("script");
+
+  if (!script || script->empty()) {
+    std::move(callback).Run(
+        CreateContentBlocksForText("Error: Missing or empty 'script' field"));
+    return;
+  }
+
+  auto wrapped_js =
+      base::StrCat({"(async function() { try {", *script,
+                    "} catch (error) { return error.toString(); } })()"});
+
+  auto* otr_profile =
+      profile_->GetPrimaryOTRProfile(true /* create_if_needed */);
+  content::WebContents::CreateParams create_params(otr_profile);
+  auto web_contents = content::WebContents::Create(create_params);
+
+  requests_.emplace_back(std::move(web_contents), wrapped_js,
+                         std::move(callback));
+
+  auto request_it = std::prev(requests_.end());
+  request_it->observer = std::make_unique<CodeSandboxWebContentsObserver>(
+      request_it->web_contents.get(),
+      base::BindOnce(&CodeExecutionTool::OnPageLoadComplete,
+                     weak_ptr_factory_.GetWeakPtr(), request_it));
+
+  request_it->web_contents->GetController().LoadURL(
+      GURL(kAIChatCodeSandboxUIURL), content::Referrer(),
+      ui::PAGE_TRANSITION_TYPED, std::string());
+
+  auto time_limit =
+      execution_time_limit_for_testing_.value_or(kExecutionTimeLimit);
+  request_it->timeout_timer.Start(
+      FROM_HERE, time_limit,
+      base::BindOnce(&CodeExecutionTool::ResolveRequest,
+                     weak_ptr_factory_.GetWeakPtr(), request_it,
+                     "Error: Time limit exceeded"));
+}
+
+void CodeExecutionTool::OnPageLoadComplete(
+    std::list<CodeExecutionRequest>::iterator request_it,
```

Comment:
You already have `CodeExecutionRequest`. Just add methods there to handle all callbacks and timers, no need to have data in `CodeExecutionRequest` and handlers in `CodeExecutionTool`.

You also shouldn't have `CodeSandboxWebContentsObserver` as a separate class.

Just try to unify everything related to a single request in one class please. You can make it have a static method like `Create` that will prepare everything and pass into a constructor.

In `CodeExecutionTool` you can have a single member with currently running requests, and a single callback that will remove those `CodeExecutionRequest` on completion.

In this setup on `CodeExecutionTool` destructor all currently active requests (with bound webcontents/timers) should be destroyed automatically.